### PR TITLE
Get target architecture from npm config environment variables

### PR DESCRIPTION
**Pull request by piranna**
Backported from https://github.com/mafintosh/prebuild/pull/156

### CODE REVIEW

**Comment by piranna** [comment]
Tests are failling for Node.js 0.10 and 0.12 due to `unzip-response` dependency. Since Node.js 0.12 will get unmaintained in just two weeks, I would remove their support at all.

**Comment by piranna** [comment]
Tests are now passing after rebase with latest master, could this be merged and published?

**Comment by mathiask88** [comment]
Is there a reason why you not added it to the array [here](https://github.com/mafintosh/prebuild-install/pull/21/files#diff-b815d9f9b45889bebd1cef9f0a873ae9L19)?

**Comment by piranna** [comment]
> Is there a reason why you not added it to the array here?

As commented at https://github.com/mafintosh/prebuild/pull/156#issuecomment-267797351:

> Problem is that the code before is adding the attributes to process.argv from the npm_config_* environment variables without removing the already defined command line arguments, and seeing the code of minimist seems that when an argument is defined several times it's converted on an Array. I don't know what's doing rc in that case, but seems it's not doing too much magic with it so it will be passed directly to the node-gyp arguments and probably break... :-/ I definitely think this can be simplified and less error prone by not doing so much magic by using instead the method I have done here.

**Comment by mathiask88** [comment]
@piranna Isn't the problem the same for all the other variables in the array? Wouldn't it be better to not add them in the loop if they already exists? Then the arguments passed directly to `prebuild-install` would be prioritized over the npm ones if I got this right.

**Comment by piranna** [comment]
> @piranna Isn't the problem the same for all the other variables in the array?

If I understood correctly the code, then yes, the other variables have the same problem too.


> Wouldn't it be better to not add them in the loop if they already exists? Then the arguments passed directly to prebuild-install would be prioritized over the npm ones if I got this right.

It would be possible to do it that way, but I also find it simpler and easier to understand to remove that loop altogether and use instead an `OR`ed assignement of the attributes, the same way it's being done with `arch` on this pull-request. I would be able to update it to do the same with the others...

**Comment by mathiask88** [comment]
I'm okay with the `||` approach. Can you update this PR for all variables?  The other loop with `process.env.npm_config_argv` should be fine I think because there are only boolean values and minimist takes the last defined value.

**Comment by piranna** [comment]
> Can you update this PR for all variables?

Done :-) Now the values are fetch from CLI arguments > values en `package.json` > npm environment variables > default values. I've also added some comments and docs explaining that.


> The other loop with process.env.npm_config_argv should be fine I think because there are only boolean values and minimist takes the last defined value.

Mostly yes, I've left them because they makes sense since they are passing `prebuild-install` arguments that were given to the `npm` command, although I would also use `OR`'ed values here too for clean-up, homogeneity and prevent failures.

Oh, and tests are passing after doing that changes, too :-D

**Comment by mathiask88** [comment]
LGTM

While reading I think we should honor the `local-address` env in the download code, but that has nothing to do with this PR :)

**Comment by piranna** [comment]
While reading I think we should honor the local-address env in the download
code

What do you mean?

**Comment by mathiask88** [comment]
Something like this https://github.com/mafintosh/prebuild-install/blob/localAddress/download.js#L64-L66

**Comment by piranna** [comment]
Yeah, that's the kind of things that should ve reviewed on the code sync.

**Comment by mathiask88** [comment]
Maybe another LGTM? @ralphtheninja @juliangruber

File: rc.js [hunk 1]
```diff
@@ -16,30 +19,34 @@ if (process.env.npm_config_argv) {
...
+ *
+ * Values are fetch from the CLI arguments, the `pkg` object (probably from the
+`* package.json` file), the environment variables set by the `npm` instance that
+ * executed `prebuild-install`, and finally some safe default values from the
+ * environment variables or the current Node.js instance
+ *
+ * @param {object} pkg
+ *
+ * @return {object} the aggregated configuration
+ */
```
Comment (juliangruber) [inline rc.js]: code style :P

Comment (piranna) [inline rc.js]: What do you mean, the jsdoc? Yeah, I forgot to document the fields of `pkg`... I can be able to add them.

Comment (piranna) [inline rc.js]: I'm seeing it, and the only field being used is `pkg.config`... Maybe we could use `config` directly instead of `pkg`?

Comment (mathiask88) [inline rc.js]: BTW the ` in L26 is misplaced :)

Comment (juliangruber) [inline rc.js]: no I meant that this project doesn't do jsdoc style comments (yet)

Comment (mathiask88) [inline rc.js]: Yeah, you are right. I'm in favor of removing the jsdoc here and maybe add full jsdoc later.

Comment (piranna) [inline rc.js]: Ok, I'm just used to it :-) I'll remove the jsdoc and just leff the comments, and later I'll open a new pull-request with the jsdoc instead.

**Comment by piranna** [comment]
Removed jsdoc and tests passing.

**Comment by piranna** [comment]
Thank you! :-D

**Comment by piranna** [comment]
When this could be published on `npm`?

**Comment by piranna** [comment]
Could this be published soon-ish? I would like to release a stable version of [libblkid](https://github.com/piranna/libblkid) with no in-development dependencies and I'm blocked due to this...

**Comment by ralphtheninja** [comment]
@piranna You know you can publish this module yourself, right? :)

**Comment by ralphtheninja** [comment]
Go nuts :)

**Comment by piranna** [comment]
> @piranna You know you can publish this module yourself, right? :)

Ok... didn't know that.. or didn't remember it :-P I don't have usually publish permissions for modules that are not mine, specially if I don't have commit permissions. Anyway, 2.1.0 published :-)

**Comment by mathiask88** [comment]
@piranna  Can u bump the repos version as well and create a release incl a tag?

**Comment by piranna** [comment]
Sorry, I don't have permissions on this repo :-(

**Comment by ralphtheninja** [comment]
> Sorry, I don't have permissions on this repo :-(

Then you probably should.

@mafintosh Help bitte! :)

**Pull request merged**